crypto/pkcs11key/key.go

```diff
@@ -249,13 +268,173 @@ func getPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkc
+	template := []*pkcs11.Attribute{
+		// CKA_EC_PARAMS contains the OID of the curve (part of the
+		// public key
+		pkcs11.NewAttribute(pkcs11.CKA_EC_PARAMS, nil),
+		// CKA_ID will allow use to find the corresponding public key
+		pkcs11.NewAttribute(pkcs11.CKA_ID, nil),
+	}
+
+	attr, err := module.GetAttributeValue(session, privateKeyHandle, template)
+	if err != nil {
+		return noKey, err
+	}
+
+	oid := []byte{}
+	id := []byte{}
+	gotOid, gotID := false, false
+	for _, a := range attr {
+		if a.Type == pkcs11.CKA_EC_PARAMS {
+			oid = a.Value
+			gotOid = true
+		} else if a.Type == pkcs11.CKA_ID {
+			id = a.Value
+			gotID = true
+		}
+	}
+	if !gotOid {
+		return noKey, errors.New("private key missing curve parameters")
+	}
+	if !gotID {
+		return noKey, errors.New("private key missing identifier (CKA_ID)")
+	}
+
+	poid := new(asn1.ObjectIdentifier)
+	asn1.Unmarshal(oid, poid)
+	curve := namedCurveFromOID(*poid)
+
+	templateSearch := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_CLASS, pkcs11.CKO_PUBLIC_KEY),
+		pkcs11.NewAttribute(pkcs11.CKA_ID, id),
+	}
+
+	if err := module.FindObjectsInit(session, templateSearch); err != nil {
+		return noKey, err
+	}
+	objs, _, err := module.FindObjects(session, 1)
+	if err != nil {
+		return noKey, err
+	}
+	if err = module.FindObjectsFinal(session); err != nil {
+		return noKey, err
+	}
+
+	if len(objs) == 0 {
+		return noKey, fmt.Errorf("public key not found")
+	}
+	publicKeyHandle := objs[0]
+
+	templatePub := []*pkcs11.Attribute{
+		pkcs11.NewAttribute(pkcs11.CKA_EC_POINT, nil),
+	}
+
+	attrPub, err := module.GetAttributeValue(session, publicKeyHandle, templatePub)
+	if err != nil {
+		return noKey, err
+	}
+
+	ecPoint := []byte{}
+	gotEcPoint := false
+	for _, a := range attrPub {
+		if a.Type == pkcs11.CKA_EC_POINT {
+			ecPoint = a.Value
+			gotEcPoint = true
+		}
+	}
+	if !gotEcPoint {
+		return noKey, errors.New("public key missing EC Point")
+	}
+
+	x, y := readECPoint(curve, ecPoint)
+	if x == nil {
+		return noKey, errors.New("invalid EC Point")
+	}
+	ecdsa := ecdsa.PublicKey{
+		Curve: curve,
+		X:     x,
+		Y:     y,
+	}
+	return ecdsa, nil
+}
+
+// Decode Q point from CKA_EC_POINT attribute
+func readECPoint(curve elliptic.Curve, ecpoint []byte) (*big.Int, *big.Int) {
+	x, y := elliptic.Unmarshal(curve, ecpoint)
+	if x == nil {
+		// http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/os/pkcs11-curr-v2.40-os.html#_ftn1
+		// PKCS#11 v2.20 specified that the CKA_EC_POINT was to be store in a DER-encoded
+		// OCTET STRING.
+		var point asn1.RawValue
+		asn1.Unmarshal(ecpoint, &point)
+		if len(point.Bytes) > 0 {
+			x, y = elliptic.Unmarshal(curve, point.Bytes)
+		}
+	}
+	return x, y
+}
+
+// Get the public key matching a private key
+func (ps *Key) loadPublicKey(module ctx, session pkcs11.SessionHandle, privateKeyHandle pkcs11.ObjectHandle) error {
+	keyType, err := getKeyType(ps.module, session, privateKeyHandle)
+	if err != nil {
+		return err
+	}
+
+	switch keyType {
```

Comment:
I believe a simpler version of this would be (since err is already declared):

```
switch keyType {
  case pkcs11.CKK_RSA:
    ps.publicKey, err = getRSAPublicKey(...)
  case pkcs11.CKK_EC:
    ps.publicKey, err = getECPublicKey(...)
  default:
    err = fmt.Errorf(...)
}
return err
```